File: Java/src/FindMissing.java
public class FindMissing {
    static int solve(int[] nums) {
        int sum = 0;
        for (int e: nums) sum += e;
        int n = nums.length;
        return n*(n+1)/2 - sum;
    }

    static int solve2(int[] nums) {
        int i = 0;
        int n = nums.length;
        while (i < n) {
            int correct = nums[i];
            if (nums[i] < n && nums[i] != nums[correct]) {
                int temp = nums[i];
                nums[i] = nums[correct];
                nums[correct] = temp;
            }
            else i++;
        }

        for (int j = 0; j < n; j++) {
            if (nums[j] != j) return j;
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] nums = {1, 3, 0};
        System.out.println(solve(nums));
        System.out.println(solve2(nums));
    }
}

File: Python/find_missing.py
def solve(nums):
    """
    given an array of integers from 0 to n, find the missing number in the array
    """
    n = len(nums)
    n = len(nums)
    return n * (n + 1) // 2 - sum(nums)


def solve2(nums):
    # using cyclic sort
    i = 0
    while i < len(nums):
        correct = nums[i]
        if nums[i] < len(nums) and nums[i] != nums[correct]:
            nums[i], nums[correct] = nums[correct], nums[i]
        else:
            i += 1

    for i in range(len(nums)):
        if nums[i] != i:
            return i
    return -1


def main():
    nums = [1, 0, 3]
    print(solve(nums))
    print(solve2(nums))


if __name__ == '__main__':
    main()
